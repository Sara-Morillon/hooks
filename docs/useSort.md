# useSort

`useSort` offers an easy way to apply local sort to a set of local data.

## Example

```tsx
import { useSort } from '@saramorillon/hooks'

interface IData {
  name: string
  age: number
}

const data: IData[] = [
  { name: 'Zola Ray', age: 30 },
  { name: 'Arlo Curtis', age: 32 },
]

function MyComponent() {
  const { state, sort } = useSort<IData>(data)

  return (
    <table>
      <thead>
        <tr>
          <th>
            Name <SortToggle dir={state.sortDir.name} onSort={(dir) => sort('name', dir)} />
          </th>
          <th>
            Age <SortToggle dir={state.sortDir.age} onSort={(dir) => sort('age', dir)} />
          </th>
        </tr>
      </thead>

      <tbody>
        {rows.map((row) => (
          <tr>
            <td>{row.name}</td>
            <td>{row.age}</td>
          </tr>
        ))}
      </tbody>
    </table>
  )
}
```

## API

```typescript
useSort<T>(data: T[], initialSort?: ISortItem<T>[]): ISorted<T>
```

### Arguments

`data: T[]` - the data that will be sorted.

`initialSort?: ISortItem<T>[]` - the initial sort, default to an empty array.

### Returns

`ISorted<T>` - an object containing:

- `rows: T[]` - the filtered, sorted and paginated rows.
- `state: ISortItem<T>[][]` - the sort state, as an ordered list.
- `sort: (field: keyof T, dir?: 'asc' | 'desc') => void` - a function to apply a sort to a field.
