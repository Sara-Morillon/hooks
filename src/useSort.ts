import { useCallback, useMemo, useState } from 'react'

export interface ISort<T> {
  column: keyof T
  dir: 'asc' | 'desc'
}

export function useSort<T>(data: T[]) {
  const [state, setState] = useState<ISort<T>[]>([])

  const sort = useCallback((column: keyof T, dir?: 'asc' | 'desc') => {
    setState((state) => {
      const newState = state.filter((s) => s.column !== column)
      if (dir) {
        newState.push({ column, dir })
      }
      return newState
    })
  }, [])

  const rows = useMemo(() => {
    return data.toSorted((rowA: T, rowB: T) => {
      for (const { column, dir } of state) {
        if (rowA[column] < rowB[column]) {
          return dir === 'asc' ? -1 : 1
        }
        if (rowA[column] > rowB[column]) {
          return dir === 'asc' ? 1 : -1
        }
      }
      return 0
    })
  }, [data, state])

  return useMemo(() => ({ state, sort, rows }), [state, sort, rows])
}
