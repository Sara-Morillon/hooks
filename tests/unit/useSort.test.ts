import { act, renderHook } from '@testing-library/react'
import { type ISortFunctions, useSort, useSortState, useSortedRows } from '../../src/useSort.js'
import { type IData, mockTableData } from '../mock.js'

describe('useSortState', () => {
  it('should return empty sort by default', () => {
    const { result } = renderHook(() => useSortState<IData>())
    expect(result.current.state).toEqual([])
  })

  it('should return initial sort', () => {
    const { result } = renderHook(() => useSortState<IData>([{ field: 'age', dir: 'desc' }]))
    expect(result.current.state).toEqual([{ field: 'age', dir: 'desc' }])
  })

  it('should change sort', () => {
    const { result } = renderHook(() => useSortState<IData>())
    act(() => result.current.sort('name', 'asc'))
    expect(result.current.state).toEqual([{ field: 'name', dir: 'asc' }])
  })

  it('should override existing sort', () => {
    const { result } = renderHook(() => useSortState<IData>())
    act(() => result.current.sort('name', 'asc'))
    act(() => result.current.sort('age', 'asc'))
    act(() => result.current.sort('name', 'desc'))
    expect(result.current.state).toEqual([
      { field: 'age', dir: 'asc' },
      { field: 'name', dir: 'desc' },
    ])
  })

  it('should clear existing sort', () => {
    const { result } = renderHook(() => useSortState<IData>())
    act(() => result.current.sort('name', 'asc'))
    act(() => result.current.sort('name'))
    expect(result.current.state).toEqual([])
  })
})

describe('useSortedRows', () => {
  it('should return non sorted rows if sort is empty', () => {
    const { result } = renderHook(() => useSortedRows(mockTableData()))
    expect(result.current).toEqual(mockTableData())
  })

  it('should return rows sorted asc', () => {
    const { result } = renderHook(() => useSortedRows(mockTableData(), [{ field: 'name', dir: 'asc' }]))
    expect(result.current).toMatchSnapshot()
  })

  it('should return rows sorted desc', () => {
    const { result } = renderHook(() => useSortedRows(mockTableData(), [{ field: 'name', dir: 'desc' }]))
    expect(result.current).toMatchSnapshot()
  })

  it('should sort by multiple fields', () => {
    const { result } = renderHook(() =>
      useSortedRows(mockTableData(), [
        { field: 'age', dir: 'asc' },
        { field: 'name', dir: 'desc' },
      ]),
    )
    expect(result.current).toMatchSnapshot()
  })

  it('should sort rows with custom sort function', () => {
    const sortFunctions: ISortFunctions<IData> = {
      address: (rowA, rowB) => rowA.address.city.localeCompare(rowB.address.city),
    }
    const { result } = renderHook(() => useSortedRows(mockTableData(), [], sortFunctions))
    expect(result.current).toMatchSnapshot()
  })
})

describe('useSort', () => {
  it('should return empty sort by default', () => {
    const { result } = renderHook(() => useSort(mockTableData()))
    expect(result.current.state).toEqual([])
  })

  it('should return initial sort', () => {
    const { result } = renderHook(() => useSort(mockTableData(), [{ field: 'age', dir: 'desc' }]))
    expect(result.current.state).toEqual([{ field: 'age', dir: 'desc' }])
  })

  it('should change sort', () => {
    const { result } = renderHook(() => useSort(mockTableData()))
    act(() => result.current.sort('name', 'asc'))
    expect(result.current.state).toEqual([{ field: 'name', dir: 'asc' }])
  })

  it('should override existing sort', () => {
    const { result } = renderHook(() => useSort(mockTableData()))
    act(() => result.current.sort('name', 'asc'))
    act(() => result.current.sort('age', 'asc'))
    act(() => result.current.sort('name', 'desc'))
    expect(result.current.state).toEqual([
      { field: 'age', dir: 'asc' },
      { field: 'name', dir: 'desc' },
    ])
  })

  it('should clear existing sort', () => {
    const { result } = renderHook(() => useSort(mockTableData()))
    act(() => result.current.sort('name', 'asc'))
    act(() => result.current.sort('name'))
    expect(result.current.state).toEqual([])
  })

  it('should return rows sorted asc', () => {
    const { result } = renderHook(() => useSort(mockTableData(), [{ field: 'name', dir: 'asc' }]))
    expect(result.current.rows).toMatchSnapshot()
  })

  it('should return rows sorted desc', () => {
    const { result } = renderHook(() => useSort(mockTableData(), [{ field: 'name', dir: 'desc' }]))
    expect(result.current.rows).toMatchSnapshot()
  })

  it('should sort by multiple fields', () => {
    const { result } = renderHook(() =>
      useSort(mockTableData(), [
        { field: 'age', dir: 'asc' },
        { field: 'name', dir: 'desc' },
      ]),
    )
    expect(result.current.rows).toMatchSnapshot()
  })

  it('should sort rows with custom sort function', () => {
    const sortFunctions: ISortFunctions<IData> = {
      address: (rowA, rowB) => rowA.address.city.localeCompare(rowB.address.city),
    }
    const { result } = renderHook(() => useSort(mockTableData(), [], sortFunctions))
    expect(result.current).toMatchSnapshot()
  })
})
