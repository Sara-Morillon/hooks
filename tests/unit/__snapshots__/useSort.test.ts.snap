// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useSort > should return rows sorted asc 1`] = `
[
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
]
`;

exports[`useSort > should return rows sorted desc 1`] = `
[
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
]
`;

exports[`useSort > should sort by multiple fields 1`] = `
[
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
]
`;

exports[`useSort > should sort rows with custom sort function 1`] = `
{
  "rows": [
    {
      "address": {
        "city": "Beckley",
        "number": 19,
        "street": "Annfield Rd",
      },
      "age": 39,
      "name": "Gerardo Pearson",
    },
    {
      "address": {
        "city": "Dowland",
        "number": 49,
        "street": "North Promenade",
      },
      "age": 31,
      "name": "Landry Pineda",
    },
    {
      "address": {
        "city": "Druimkinnerras",
        "number": 47,
        "street": "Crescent Avenue",
      },
      "age": 30,
      "name": "Zola Ray",
    },
    {
      "address": {
        "city": "Frostenden",
        "number": 13,
        "street": "Sandyhill Rd",
      },
      "age": 37,
      "name": "Annabella Vargas",
    },
    {
      "address": {
        "city": "Gartavaich",
        "number": 11,
        "street": "Guildry Street",
      },
      "age": 30,
      "name": "Sterling Pugh",
    },
    {
      "address": {
        "city": "Haslington",
        "number": 12,
        "street": "Warner Close",
      },
      "age": 32,
      "name": "Deandre Vu",
    },
    {
      "address": {
        "city": "Kirtling Green",
        "number": 6,
        "street": "Witney Way",
      },
      "age": 40,
      "name": "Diego Fitzpatrick",
    },
    {
      "address": {
        "city": "Little Woolstone",
        "number": 6,
        "street": "Overton Circle",
      },
      "age": 39,
      "name": "Alexis West",
    },
    {
      "address": {
        "city": "Lynemore",
        "number": 31,
        "street": "South Crescent",
      },
      "age": 35,
      "name": "Stefan Hart",
    },
    {
      "address": {
        "city": "Malling",
        "number": 96,
        "street": "Cunnery Rd",
      },
      "age": 34,
      "name": "Kiara Nava",
    },
    {
      "address": {
        "city": "Myddfai",
        "number": 19,
        "street": "Ilchester Road",
      },
      "age": 38,
      "name": "Gemma Terrell",
    },
    {
      "address": {
        "city": "New Inn",
        "number": 50,
        "street": "Iolaire Road",
      },
      "age": 32,
      "name": "Arlo Curtis",
    },
    {
      "address": {
        "city": "Parkham",
        "number": 19,
        "street": "Well Lane",
      },
      "age": 30,
      "name": "Olivia Camacho",
    },
    {
      "address": {
        "city": "Rowley",
        "number": 67,
        "street": "Harehills Lane",
      },
      "age": 40,
      "name": "Alaya Winters",
    },
    {
      "address": {
        "city": "Sandaig",
        "number": 97,
        "street": "Wade Lane",
      },
      "age": 38,
      "name": "Kimora Higgins",
    },
    {
      "address": {
        "city": "Sopworth",
        "number": 41,
        "street": "Hounslow Rd",
      },
      "age": 34,
      "name": "Ryker Gregory",
    },
    {
      "address": {
        "city": "Winslade",
        "number": 42,
        "street": "Folkestone Road",
      },
      "age": 35,
      "name": "Jaxen Smith",
    },
  ],
  "sort": [Function],
  "state": [
    {
      "dir": "asc",
      "field": "address",
    },
  ],
}
`;

exports[`useSortedRows > should return rows sorted asc 1`] = `
[
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
]
`;

exports[`useSortedRows > should return rows sorted desc 1`] = `
[
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
]
`;

exports[`useSortedRows > should sort by multiple fields 1`] = `
[
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
]
`;

exports[`useSortedRows > should sort rows with custom sort function 1`] = `
[
  {
    "address": {
      "city": "Beckley",
      "number": 19,
      "street": "Annfield Rd",
    },
    "age": 39,
    "name": "Gerardo Pearson",
  },
  {
    "address": {
      "city": "Dowland",
      "number": 49,
      "street": "North Promenade",
    },
    "age": 31,
    "name": "Landry Pineda",
  },
  {
    "address": {
      "city": "Druimkinnerras",
      "number": 47,
      "street": "Crescent Avenue",
    },
    "age": 30,
    "name": "Zola Ray",
  },
  {
    "address": {
      "city": "Frostenden",
      "number": 13,
      "street": "Sandyhill Rd",
    },
    "age": 37,
    "name": "Annabella Vargas",
  },
  {
    "address": {
      "city": "Gartavaich",
      "number": 11,
      "street": "Guildry Street",
    },
    "age": 30,
    "name": "Sterling Pugh",
  },
  {
    "address": {
      "city": "Haslington",
      "number": 12,
      "street": "Warner Close",
    },
    "age": 32,
    "name": "Deandre Vu",
  },
  {
    "address": {
      "city": "Kirtling Green",
      "number": 6,
      "street": "Witney Way",
    },
    "age": 40,
    "name": "Diego Fitzpatrick",
  },
  {
    "address": {
      "city": "Little Woolstone",
      "number": 6,
      "street": "Overton Circle",
    },
    "age": 39,
    "name": "Alexis West",
  },
  {
    "address": {
      "city": "Lynemore",
      "number": 31,
      "street": "South Crescent",
    },
    "age": 35,
    "name": "Stefan Hart",
  },
  {
    "address": {
      "city": "Malling",
      "number": 96,
      "street": "Cunnery Rd",
    },
    "age": 34,
    "name": "Kiara Nava",
  },
  {
    "address": {
      "city": "Myddfai",
      "number": 19,
      "street": "Ilchester Road",
    },
    "age": 38,
    "name": "Gemma Terrell",
  },
  {
    "address": {
      "city": "New Inn",
      "number": 50,
      "street": "Iolaire Road",
    },
    "age": 32,
    "name": "Arlo Curtis",
  },
  {
    "address": {
      "city": "Parkham",
      "number": 19,
      "street": "Well Lane",
    },
    "age": 30,
    "name": "Olivia Camacho",
  },
  {
    "address": {
      "city": "Rowley",
      "number": 67,
      "street": "Harehills Lane",
    },
    "age": 40,
    "name": "Alaya Winters",
  },
  {
    "address": {
      "city": "Sandaig",
      "number": 97,
      "street": "Wade Lane",
    },
    "age": 38,
    "name": "Kimora Higgins",
  },
  {
    "address": {
      "city": "Sopworth",
      "number": 41,
      "street": "Hounslow Rd",
    },
    "age": 34,
    "name": "Ryker Gregory",
  },
  {
    "address": {
      "city": "Winslade",
      "number": 42,
      "street": "Folkestone Road",
    },
    "age": 35,
    "name": "Jaxen Smith",
  },
]
`;
